libraries/rush-lib/src/logic/operations/CacheableOperationPlugin.ts

```diff
@@ -0,0 +1,875 @@
+
+  private _applyShellOperationRunner(runner: ShellOperationRunner, context: ICreateOperationsContext): void {
+    const {
+      buildCacheConfiguration,
+      cobuildConfiguration,
+      phaseSelection: selectedPhases,
+      projectChangeAnalyzer
+    } = context;
+    const { hooks } = runner;
+
+    const buildCacheContext: IOperationBuildCacheContext = this._getBuildCacheContextByRunnerOrThrow(runner);
+
+    hooks.beforeExecute.tapPromise(
+      PLUGIN_NAME,
+      async (beforeExecuteContext: IOperationRunnerBeforeExecuteContext) => {
+        const earlyReturnStatus: OperationStatus | undefined = await (async () => {
+          const {
+            context,
+            runner,
+            terminal,
+            lastProjectDeps,
+            projectDeps,
+            trackedProjectFiles,
+            logPath,
+            errorLogPath,
+            rushProject,
+            phase,
+            commandName,
+            commandToRun,
+            earlyReturnStatus
+          } = beforeExecuteContext;
+          if (earlyReturnStatus) {
+            // If there is existing early return status, we don't need to do anything
+            return earlyReturnStatus;
+          }
+
+          if (!projectDeps && buildCacheContext.isSkipAllowed) {
+            // To test this code path:
+            // Remove the `.git` folder then run "rush build --verbose"
+            terminal.writeLine({
+              text: PrintUtilities.wrapWords(
+                'This workspace does not appear to be tracked by Git. ' +
+                  'Rush will proceed without incremental execution, caching, and change detection.'
+              ),
+              foregroundColor: ColorValue.Cyan
+            });
+          }
+
+          const buildCacheTerminal: ITerminal = this._getBuildCacheTerminal({
+            runner,
+            buildCacheConfiguration,
+            rushProject,
+            collatedWriter: context.collatedWriter,
+            logFilenameIdentifier: runner.logFilenameIdentifier,
+            quietMode: context.quietMode,
+            debugMode: context.debugMode
+          });
+          buildCacheContext.buildCacheTerminal = buildCacheTerminal;
+
+          let projectBuildCache: ProjectBuildCache | undefined = await this._tryGetProjectBuildCacheAsync({
+            buildCacheConfiguration,
+            runner,
+            rushProject,
+            phase,
+            selectedPhases,
+            projectChangeAnalyzer,
+            commandName,
+            commandToRun,
+            terminal: buildCacheTerminal,
+            trackedProjectFiles,
+            operationMetadataManager: context._operationMetadataManager
+          });
+
+          // Try to acquire the cobuild lock
+          let cobuildLock: CobuildLock | undefined;
+          if (cobuildConfiguration?.cobuildEnabled) {
+            if (
+              cobuildConfiguration?.cobuildLeafProjectLogOnlyAllowed &&
+              rushProject.consumingProjects.size === 0 &&
```

Comment:
```suggestion
              runner.consumers.size === 0 &&
```
In a repository with multiple phases, the _Operation_ needs to have no consumers for this to work, not just the Rush project.